### ISSUE TITLE
Fine-Tune or Adapt Models on Domain Data

### ISSUE THREAD

**Issue by toddllm** [toddllm/luanti-voyager#29]
### Overview
To boost the agent's proficiency, fine-tune the language model on data specific to the Luanti voxel world. This involves gathering transcripts of successful episodes and game-relevant data, then using modern fine-tuning techniques to create a domain expert model.

### Inspiration from AI Makerspace
Makerspace courses showed that targeted fine-tuning and continued pre-training can significantly improve a model's performance in-domain. They emphasize leveraging fine-tuning to go from a general foundation model to a *domain expert* model.

**Relevant AI Makerspace Resources:**
- Fine-Tuning workshops
- PEFT and LoRA techniques
- Domain Adaptation sessions
- [Awesome AIM Index](https://github.com/AI-Maker-Space/Awesome-AIM-Index)

#### 1. Data Collection System
```python
class AgentDataCollector:
    def __init__(self):
        self.episode_buffer = []
        self.success_threshold = 0.8
        
    def record_episode(self, agent_id: str):
        """Record full episode for training data"""
        episode = {
            "id": f"{agent_id}_{timestamp}",
            "conversations": [],
            "outcomes": {},
            "success_score": 0.0
        }
        
        # Collect all prompts and responses
        for turn in agent.history:
            episode["conversations"].append({
                "prompt": turn.prompt,
                "response": turn.response,
                "context": turn.world_state,
                "action_result": turn.result
            })
        
        # Evaluate episode success
        episode["success_score"] = self.evaluate_episode(episode)
        
        if episode["success_score"] > self.success_threshold:
            self.episode_buffer.append(episode)
```

#### 2. Dataset Preparation
```python
def prepare_finetuning_dataset(episodes):
    """Convert episodes to training format"""
    dataset = []
    
    for episode in episodes:
        for conv in episode["conversations"]:
            # Format for instruction tuning
            sample = {
                "instruction": format_system_prompt(conv["context"]),
                "input": conv["prompt"],
                "output": conv["response"],
                "metadata": {
                    "success": conv["action_result"]["success"],
                    "world_state": conv["context"]
                }
            }
            
            # Only include successful actions
            if sample["metadata"]["success"]:
                dataset.append(sample)
    
    return dataset
```

#### 3. LoRA Fine-Tuning
```python
from peft import LoraConfig, get_peft_model, TaskType
from transformers import AutoModelForCausalLM, TrainingArguments

# LoRA configuration for efficient fine-tuning
peft_config = LoraConfig(
    task_type=TaskType.CAUSAL_LM,
    inference_mode=False,
    r=16,  # Rank
    lora_alpha=32,
    lora_dropout=0.1,
    target_modules=["q_proj", "v_proj", "k_proj", "o_proj"]
)

# Load base model
base_model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-2-7b-hf",
    torch_dtype=torch.float16,
    device_map="auto"
)

# Apply LoRA
model = get_peft_model(base_model, peft_config)

# Training arguments
training_args = TrainingArguments(
    output_dir="./luanti-voyager-lora",
    num_train_epochs=3,
    per_device_train_batch_size=4,
    gradient_accumulation_steps=4,
    warmup_steps=100,
    logging_steps=25,
    save_strategy="epoch",
    evaluation_strategy="epoch",
    learning_rate=2e-4,
    fp16=True,
    push_to_hub=True
)
```

### Training Data Categories

1. **Successful Episodes**
   - Complete survival scenarios
   - Efficient building sequences
   - Creative problem solutions

2. **Domain Knowledge**
   - Crafting recipes with context
   - Biome-specific strategies
   - Combat techniques

3. **Error Recovery**
   - How to escape dangerous situations
   - Learning from failures
   - Alternative approaches

4. **Multi-Agent Coordination**
   - Communication patterns
   - Task delegation
   - Collaborative building

### Curriculum Learning Approach
```python
class CurriculumTrainer:
    def __init__(self):
        self.difficulty_levels = [
            "basic_survival",     # Simple tasks
            "resource_gathering", # Intermediate
            "complex_building",   # Advanced
            "multi_agent_coord"   # Expert
        ]
    
    def train_progressive(self, model, datasets):
        """Train model progressively on harder tasks"""
        for level in self.difficulty_levels:
            print(f"Training on {level} data...")
            
            # Fine-tune on current level
            trainer = Trainer(
                model=model,
                train_dataset=datasets[level],
                args=training_args
            )
            trainer.train()
            
            # Evaluate before moving to next level
            if not self.meets_performance_threshold(model, level):
                print(f"Additional training needed for {level}")
                trainer.train()  # Additional epoch
```

### Continuous Learning
```python
class ContinuousLearning:
    def __init__(self, model_path):
        self.model = self.load_model(model_path)
        self.experience_buffer = ExperienceReplay(max_size=10000)
        
    def update_from_experience(self):
        """Periodically update model with new experiences"""
        if len(self.experience_buffer) > 1000:
            # Create mini-dataset from recent successes
            dataset = self.prepare_dataset(
                self.experience_buffer.get_successful_episodes()
            )
            
            # Quick fine-tune (few steps)
            self.quick_finetune(dataset, steps=50)
            
            # Save checkpoint
            self.save_checkpoint()
```

### Benefits
- Domain-specific expertise baked into model
- Reduced reliance on lengthy prompts
- Faster and more accurate decisions
- Learns from collective agent experiences
- Adapts to player's world over time

### Acceptance Criteria
- [ ] Data collection pipeline for episodes
- [ ] Dataset preparation tools
- [ ] LoRA fine-tuning implementation
- [ ] Training script with wandb logging
- [ ] Model evaluation on game tasks
- [ ] Comparison with base model performance
- [ ] Documentation for training custom models
- [ ] Model sharing via HuggingFace Hub

### Evaluation Metrics
- Task completion rate improvement
- Average survival time increase
- Reduction in invalid actions
- Faster goal achievement
- Lower token usage per decision

**Labels:** `enhancement`, `ai-makerspace-inspired`, `fine-tuning`, `machine-learning`

**Comment by toddllm** [comment]
## AI Makerspace Resources Update

### 🎓 Specific AI Makerspace Sessions

**Continued Pretraining and Fine-Tuning with Unsloth**
- 📓 **[Continued Pretraining Notebook](https://colab.research.google.com/drive/1oqd8PrUx_6NPlNCP2TLxMYn_z44rNrM1)**
- 📓 **[Fine-Tuning Notebook](https://colab.research.google.com/drive/1UHCo6cHQmCpmbdgZIx5qI0BeFEw8lgpX)**

**Quantization and QLoRA**
- Covered in the optimization notebooks with practical fine-tuning examples

### Key Implementation Insights from Unsloth Notebooks

#### 1. Efficient Data Collection Pipeline (from the notebooks)
The notebooks show how to structure training data:

```python
# From the Unsloth fine-tuning notebook
def format_game_episode(episode):
    """Convert game episodes to training format"""
    formatted_data = []
    
    for turn in episode:
        # Alpaca-style formatting for instruction tuning
        sample = {
            "instruction": f"""You are a Luanti game agent.
World state: {turn['world_state']}
Goal: {turn['goal']}
What action should you take?""",
            
            "input": turn['observation'],
            
            "output": turn['action_taken'],
            
            # Include success as metadata
            "metadata": {
                "success": turn['success'],
                "reward": turn.get('reward', 0)
            }
        }
        
        # Only include successful actions for initial training
        if sample['metadata']['success']:
            formatted_data.append(sample)
    
    return formatted_data
```

#### 2. Unsloth Memory-Efficient Setup (from the notebook)
The notebook demonstrates 4x less memory usage:

```python
# Directly from the Unsloth notebook
from unsloth import FastLanguageModel

# Load model with Unsloth optimizations
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name="unsloth/llama-3-8b-bnb-4bit",
    max_seq_length=2048,  # Reasonable for game contexts
    dtype=None,  # Auto-detect
    load_in_4bit=True,  # Use 4-bit quantization
)

# Add LoRA adapters with Unsloth
model = FastLanguageModel.get_peft_model(
    model,
    r=16,  # Rank - higher = more capacity
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj",
                    "gate_proj", "up_proj", "down_proj"],
    lora_alpha=16,
    lora_dropout=0,  # Optimized for Unsloth
    bias="none",
    use_gradient_checkpointing="unsloth",  # Special optimization
    random_state=3407,
)
```

#### 3. Game-Specific Training Strategy (adapted from notebooks)
Implement curriculum learning for game agents:

```python
# Progressive difficulty training
training_stages = {
    "stage1_survival": {
        "data": "basic_survival_episodes.json",
        "epochs": 3,
        "learning_rate": 2e-4,
        "focus": "Don't die, find shelter"
    },
    "stage2_crafting": {
        "data": "crafting_episodes.json", 
        "epochs": 2,
        "learning_rate": 1e-4,
        "focus": "Learn recipes and tool use"
    },
    "stage3_building": {
        "data": "construction_episodes.json",
        "epochs": 2,
        "learning_rate": 5e-5,
        "focus": "Complex multi-step building"
    }
}

# Train progressively
for stage_name, config in training_stages.items():
    print(f"Training {stage_name}...")
    
    dataset = load_dataset("json", data_files=config["data"])
    
    trainer = SFTTrainer(
        model=model,
        tokenizer=tokenizer,
        train_dataset=dataset,
        dataset_text_field="text",
        max_seq_length=2048,
        dataset_num_proc=2,
        packing=True,  # Unsloth optimization
        args=TrainingArguments(
            per_device_train_batch_size=2,
            gradient_accumulation_steps=4,
            warmup_steps=5,
            num_train_epochs=config["epochs"],
            learning_rate=config["learning_rate"],
            fp16=not torch.cuda.is_bf16_supported(),
            bf16=torch.cuda.is_bf16_supported(),
            logging_steps=1,
            optim="adamw_8bit",  # Memory efficient
            weight_decay=0.01,
            lr_scheduler_type="linear",
            output_dir=f"outputs_{stage_name}",
        ),
    )
    
    trainer.train()
```

#### 4. Continuous Learning Implementation (from the patterns)
Update model with new experiences:

```python
# Periodic model updates with new data
class ContinuousFineTuner:
    def __init__(self, base_model_path):
        self.model, self.tokenizer = FastLanguageModel.from_pretrained(
            base_model_path,
            max_seq_length=2048,
            load_in_4bit=True,
        )
        self.experience_buffer = []
        
    def add_episode(self, episode):
        """Add successful episode to buffer"""
        if episode['success_rate'] > 0.8:
            formatted = format_game_episode(episode)
            self.experience_buffer.extend(formatted)
    
    def update_model(self):
        """Quick fine-tune on recent experiences"""
        if len(self.experience_buffer) < 100:
            return
            
        # Create temporary dataset
        dataset = Dataset.from_list(self.experience_buffer[-500:])
        
        # Quick fine-tune (few steps)
        FastLanguageModel.for_training(self.model)
        
        trainer = SFTTrainer(
            model=self.model,
            train_dataset=dataset,
            max_seq_length=2048,
            args=TrainingArguments(
                num_train_epochs=1,
                per_device_train_batch_size=4,
                learning_rate=5e-5,
                max_steps=50,  # Just a quick update
                output_dir="continuous_updates",
            ),
        )
        
        trainer.train()
        
        # Clear old experiences
        self.experience_buffer = self.experience_buffer[-200:]
```

#### 5. Export and Deployment (from the notebook)
Save and share fine-tuned models:

```python
# Save with Unsloth optimizations
model.save_pretrained_merged(
    "luanti-voyager-llama3-8b",
    tokenizer,
    save_method="merged_16bit",  # Full precision merge
)

# Or save just the LoRA adapters
model.save_pretrained_gguf(
    "luanti-voyager-qlora",
    tokenizer,
    quantization_method="q4_k_m"  # For llama.cpp
)

# Push to HuggingFace Hub
model.push_to_hub_merged(
    "username/luanti-voyager-agent",
    tokenizer,
    save_method="lora",  # Just the adapter
    token="hf_..."
)
```

### Practical Training Tips from AI Makerspace

1. **Start with QLoRA** - Uses only 4-6GB VRAM for 7B models
2. **Use Unsloth** - 2-4x faster training with same quality
3. **Collect Quality Data** - 1000 good examples > 10000 mediocre ones
4. **Progressive Training** - Start simple, increase complexity
5. **Monitor Overfitting** - Game scenarios are limited, watch for memorization

The notebooks emphasize that fine-tuning isn't just about running training—it's about carefully curating data and progressively building agent capabilities.
